workflow/controller/controller.go

```diff
@@ -534,6 +534,65 @@ func (wfc *WorkflowController) signalContainers(namespace string, podName string
 	return time.Duration(*pod.Spec.TerminationGracePeriodSeconds) * time.Second, nil
 }
 
+func (wfc *WorkflowController) deleteOffloadedNodes() error {
+	oldRecords, err := wfc.offloadNodeStatusRepo.ListOldOffloads(wfc.GetManagedNamespace())
+	if err != nil {
+		return err
+	}
+	if len(oldRecords) == 0 {
+		log.Info("Zero old offloads, nothing to do")
+		return nil
+	}
+	workflows, err := util.NewWorkflowLister(wfc.wfInformer).List()
+	if err != nil {
+		return err
+	}
+	for _, wf := range workflows {
+		log.WithField("uid", wf.UID).Debug("Locking workflow")
+		key, err := cache.MetaNamespaceKeyFunc(wf)
+		if err != nil {
+			return err
+		}
+		wfc.workflowKeyLock.Lock(key)
+	}
+	defer func() {
```

Comment:
this will lock all the workflows, potentially locking the controller and preventing reconciliation - this is a show stopper - instead create a new func (e.g. named deleteOffloadedNodesForWorkflow()) and lock/defer at the start of the func